test/slim_check.lean

```diff
@@ -0,0 +1,75 @@
+
+import system.random.basic data.nat.prime
+
+instance fin.has_one' {n} [fact (0 < n)] : has_one (fin n) :=
+⟨ fin.of_nat' 1 ⟩
+
+instance fin.has_pow' {n} [fact (0 < n)] : has_pow (fin n) ℕ :=
+⟨ monoid.pow ⟩
+
+/-- fermat's primality test -/
+def primality_test (p : ℕ) (h : fact (0 < p)) : rand bool :=
+if h : 2 ≤ p-1 then do
+  n ← rand.random_r 2 (p-1), -- `random_r` requires a proof of `2 ≤ p-1` but it is dischared using `assumption`
+  return $ (fin.of_nat' n : fin p)^(p-1) = 1 -- we do arithmetic with `fin n` so that modulo and multiplication are interleaved
+else return (p = 2)
+
+def iterated_primality_test_aux (p : ℕ) (h : fact (0 < p)) : ℕ → rand bool
+| 0 := pure tt
+| (n+1) := do
+  b ← primality_test p h,
+  if b
+    then iterated_primality_test_aux n
+    else pure ff
+
+def iterated_primality_test (p : ℕ) : rand bool :=
+if h : 0 < p
+  then iterated_primality_test_aux p h 10
+  else pure ff
+
+def mfirst {α} {m} [monad m] (f : α → m bool) : list α → m (option α)
+| [] := pure none
+| (x :: xs) :=
+mcond (f x)
+  (pure x)
+  (mfirst xs)
+
+def find_prime_aux (p : ℕ) (h : 1 ≤ p / 2) : ℕ → rand (option ℕ)
+| 0 := pure none
+| (n+1) := do
+  k ← rand.random_r 1 (p / 2),
+  let xs := (list.range' k 20).map (λ i, 2*i+1),
+  some r ← mfirst iterated_primality_test xs
+    | find_prime_aux n,
+  pure r
```

Comment:
Maybe say that `n` here a bound on how long we search for?

Reply:
Why are you passing `h`? Isn't it unused?

Reply:
Oh, sorry, it's being silently consumed by `rand.random_r`. Confusing?

Reply:
Yeah, it could be confusing. I was hoping it would save me the writing of overly long proof terms as arguments but it hasn't so far. It can also be achieved on the caller side with other means (like defining special notation to build a term using a tactic like `simp` or `norm_num`). I can get rid of this device if you think that will be clearer

Reply:
Yeah, it seems it may be more trouble than it's worth.

Reply:
FWIW, I prefer seeing that random wants the proof, tbh. implicit can be a bit magic.